lib/ttfunk/table/cff/dict.rb

```diff
@@ -0,0 +1,264 @@
+require 'bigdecimal'
+
+module TTFunk
+  class Table
+    class Cff < TTFunk::Table
+      class Dict < TTFunk::SubTable
+        class InvalidOperandError < StandardError; end
+        class TooManyOperandsError < StandardError; end
+
+        # for regular single-byte operators
+        OPERATOR_BZERO = 0..21
+        OPERAND_BZERO = [28..30, 32..254].freeze
+
+        # for operators that are two bytes wide
+        WIDE_OPERATOR_BZERO = 12
+        WIDE_OPERATOR_ADJUSTMENT = 1200
+
+        # maximum number of operands allowed per operator
+        MAX_OPERANDS = 48
+
+        # used to validate operands expressed in scientific notation
+        VALID_SCI_SIGNIFICAND_RE = /\A-?\d+(?:\.\d+)?\z/
+        VALID_SCI_EXPONENT_RE = /\A-?\d+\z/
+
+        include Enumerable
+
+        def [](operator)
+          @dict[operator]
+        end
+
+        def each(&block)
+          @dict.each(&block)
+        end
+
+        alias each_pair each
+
+        def encode
+          ''.tap do |result|
```

Comment:
Just heads up: this pattern will have to be replaced. You may have seen that Prawan master is updated to use Ruby 2.3. That entails frozen strings as we're trying to stay close to community style guide (and coincidentally Rubocop's default configuration). I haven't updated TTFunk yet to reflect the change but I will before the release. You can keep using string mutation trough out these PRs. I will handle the update after the merge. Or you can help me a bit and opt for immutable strings in your PRs.

I'd suggest `.map{}.join('')` as a replacement pattern here.

Reply:
According to a couple blog articles, we can also just do `''.dup.tap do ...` also. What do you think about that?

Reply:
I mean, technically you can use `+''` to get a mutable string, which is more concise and intention revealing. But why use mutation if there's an equally good alternative?

Reply:
I was just thinking `<<` is more efficient than allocating a temporary array.

Reply:
I thought that they should be comparable in performance since  `Array#join` internally is very similar to `s = ''; array.each {|x| s << x }`. `.map.join` just looks a bit more natural to my eye. But I got curious and decided to check.

```ruby
require 'benchmark/ips'

$a = (0..1_000).map(&:to_s)

Benchmark.ips do |x|
  x.report "String#<<" do
    s = +''
    $a.each do |n|
      s << n
    end
  end

  x.report "Array#join" do
    $a.join('')
  end
end
```

```
 String#<<     13.332k (±16.5%) i/s -     64.183k in   5.062950s
Array#join     17.965k (±17.9%) i/s -     85.050k in   5.069048s
```

Apparently, `.join` is about 30% faster.

Reply:
Ok, but the construction of the array should be part of the benchmark. The individual `<<` operations vs a single `join` (which I'm sure performs a loop in C land) might be fast, but constructing the array is an operation that we are introducing to the algorithm, so it should be measured:

```ruby
require 'benchmark/ips'

Benchmark.ips do |x|
  x.report "String#<<" do
    s = +''
    1_000.times do |i|
      s << i.to_s
    end
  end

  x.report "Array#join" do
    a = (0..1_000).map(&:to_s)
    a.join('')
  end

  x.compare!
end
```

As it turns out, the two techniques result in about the same i/s:

```
Warming up --------------------------------------
           String#<<   545.000  i/100ms
          Array#join   490.000  i/100ms
Calculating -------------------------------------
           String#<<      5.655k (± 4.3%) i/s -     28.340k in   5.020904s
          Array#join      5.197k (± 4.3%) i/s -     25.970k in   5.006597s

Comparison:
           String#<<:     5655.1 i/s
          Array#join:     5196.7 i/s - same-ish: difference falls within error
```